### PR TITLE
Mirror data fixes

### PR DESCRIPTION
**Pull request by fschulze**
This fixes the stored data in mirror indexes, so it can all be referenced and properly deleted.

### CODE REVIEW

File: server/devpi_server/extpypi.py [hunk 1]
```diff
@@ -157,6 +159,13 @@ def modify(self, index=None, **kw):
             self.ixconfig = newconfig
             return newconfig
 
+    def add_project(self, project):
```
Comment (gotcha) [inline server/devpi_server/extpypi.py]: Rename function to ``update_project_names`` ?

File: server/devpi_server/extpypi.py [hunk 2]
```diff
@@ -245,6 +254,9 @@ def _save_cache_links(self, project, links, serial):
         if old != data:
             threadlog.debug("saving changed simplelinks for %s: %s", project, data)
             key.set(data)
+            # add project if it isn't already, so deletion and offline mode
```
Comment (gotcha) [inline server/devpi_server/extpypi.py]: Rephrase comment:

maintain list of currently cached project names to enable deletion and offline mode

File: server/devpi_server/model.py [hunk 3]
```diff
@@ -679,6 +679,10 @@ def _set_versiondata(self, metadata):
         if version not in versions:
             versions.add(version)
             self.key_projversions(project).set(versions)
+        self.add_project(project)
+
+    def add_project(self, project):
```
Comment (gotcha) [inline server/devpi_server/model.py]: Same as above, rename function to ``update_project_names`` ?

File: server/devpi_server/extpypi.py [hunk 4]
```diff
@@ -216,17 +217,18 @@ def _get_remote_projects(self):
 
     def list_projects_perstage(self):
         """ return set of all projects served through the mirror. """
+        if self.offline:
+            threadlog.warn("using stale projects list")
```
Comment (gotcha) [inline server/devpi_server/extpypi.py]: Would it make sense to log ``offline mode: using stale projects list`` ?

File: server/devpi_server/extpypi.py [hunk 5]
```diff
@@ -216,17 +217,18 @@ def _get_remote_projects(self):
...
         if self.cache_projectnames.is_fresh(self.cache_expiry):
             projects = self.cache_projectnames.get()
         else:
             # no fresh projects or None at all, let's go remote
             try:
                 projects = self._get_remote_projects()
             except self.UpstreamError:
-                if not self.cache_projectnames.exists():
-                    raise
                 threadlog.warn("using stale projects list")
```
Comment (gotcha) [inline server/devpi_server/extpypi.py]: And express here what type of error happened...

File: server/devpi_server/extpypi.py [hunk 6]
```diff
@@ -298,7 +300,9 @@ def get_simplelinks_perstage(self, project):
         """
         project = normalize_name(project)
         is_fresh, links, cache_serial = self._load_cache_links(project)
-        if is_fresh:
+        if is_fresh or self.offline:
```
Comment (gotcha) [inline server/devpi_server/extpypi.py]: I do not understand test.offline being tested twice.
This might ne to be refactored to better express intent.

Comment (fschulze) [reply server/devpi_server/extpypi.py]: This is changed in the 404 handling commit later on

**Comment by fschulze** [comment]
I've made a final rebase to squash related commits.
I also added a new UpstreamNotFoundError exception class instead of using HTTPNotFound in the last commit about 404 handling.

**Comment by fschulze** [comment]
I also did the is_fresh -> is_expired refactoring and added caching of 404s which makes this code much closer to before in behaviour.

**Pull request merged**